### PR TITLE
Cleanup Renderer API to Remove Fixed-Function State

### PR DESCRIPTION
**Pull request by Web-eWorks**
Matrix state ~~must~~ did die!

Remove all OpenGL 2.x like matrix state methods from the renderer API, now the matrix state is the responsibility of the caller via MatrixTicket / StateTicket.

Still TODO is improving the viewport state tracking API, although that should be rather simple to fix via promoting the Viewport struct to the top-level Renderer API.

Needs a little dead-code-elimination because I wrote this between the hours of 3 and 4 AM, but should produce identical results to the previous system.

Ping @fluffyfreak - I recognize that `XXX` comment annotation and this was actually surprisingly simple to do, so there's gotta be a bug in it somewhere, right? *Right?*

### CODE REVIEW

**Comment by Web-eWorks** [comment]
For the record, I'm not 100% happy with this solution as we're still keeping some state in the renderer, but it's still miles better than retaining unnecessary fixed-function semantics in our API. I'll revisit this when I get around to replacing individual uniforms with UBOs, as that will provide a decent opportunity to remove this kind of state "for free".

File: src/graphics/opengl/RendererGL.cpp
Comment (fluffyfreak) [inline src/graphics/opengl/RendererGL.cpp:601]: Remove instead of comment out

File: src/graphics/opengl/RendererGL.cpp
Comment (fluffyfreak) [inline src/graphics/opengl/RendererGL.cpp:709]: Remove instead of comment out

File: src/graphics/Renderer.h
Comment (fluffyfreak) [inline src/graphics/Renderer.h:153]: Deprecated?

**Comment by fluffyfreak** [comment]
It looks good and I agree with the direction 👍
When I wrote the Renderer state tracking stuff it was to avoid the OpenGL calls querying state, which are a performance killer especially on ATi/AMD drivers. 
It was never really with a view to refactoring it like this so this is good progress!

File: src/graphics/opengl/RendererGL.cpp
Comment (Web-eWorks) [reply src/graphics/opengl/RendererGL.cpp:601]: This is that dead code elimination I was talking about - forgot to clean this up, but I still have one more commit to go, I'll clean it up then.

File: src/graphics/Renderer.h
Comment (Web-eWorks) [reply src/graphics/Renderer.h:153]: I deprecated the existing methods (to see the extent of their use) before removing them, this is an unintentional slip.

**Comment by Web-eWorks** [comment]
While I'm at it, I'm adding explicit type conversion between the `float` and `double` versions of `matrix4x4` and completely removing the "helper function" that was previously involved. I'll also be removing the `double` overload of `Renderer::SetTransform` and fixing things up so they communicate better to the reader that there's a loss of precision involved in passing double-based matricies to the renderer.

**Comment by Web-eWorks** [comment]
Fixed up Viewport handling, removed the `matrix4x4d` overload of `Renderer::SetTransform` (it makes it a little more messy when calling, but also explicitly clarifies that you're losing precision), and ran some clang-format over files that have slipped out of format compliance.

This also has the side-effect of majorly cleaning up `GasGiantJobs.cpp`, as we no longer need twelve lines to enter/leave orthographic state and we also no longer stomp renderer state!

**Comment by Web-eWorks** [comment]
@fluffyfreak let me know if you want to review the new changes - if you're swamped currently I'm good to go ahead and merge, but I wanted to make sure I got your sign-off first!

File: src/core/GuiApplication.cpp
Comment (fluffyfreak) [inline src/core/GuiApplication.cpp:28]: I need to test this kind of initialisation with Visual Studio builds, and non GCC builds basically.

**Comment by fluffyfreak** [comment]
I'm all good with this conceptually, I just need to check that it builds in Visual Studio.
I don't have time to do so tonight, but hopefully will tomorrow.

File: src/core/GuiApplication.cpp
Comment (Web-eWorks) [reply src/core/GuiApplication.cpp:28]: This *should* be standard-compliant list-initialization, but don't quote me on that. Appveyor didn't build because there's a merge conflict but I don't think it had problems with it.

Returning a list-initialized object (and this particular call) is a form of `copy-list-initialization`, which as far as I can tell is directly standard-compliant according to [the C++ reference](https://en.cppreference.com/w/cpp/language/list_initialization).

**Comment by laarmen** [comment]
Yeah I don't think Andrew had any doubt about the code being standard
compliant, but more about VS knowing about the standard ;-)

That said, they've improved dramatically since the days of C++11.

File: src/core/GuiApplication.cpp
Comment (fluffyfreak) [reply src/core/GuiApplication.cpp:28]: Yeah this isn't your fault or a C++ standards issue 😉 it's a does-VS-conform-to-the-C++-standards issue, which with this particular list initialisation I have seen it fail before, so I want to double check it before it's merged

**Review by fluffyfreak — APPROVED** [review]
This is all good from my perspective 👍

**Pull request merged**
